newton/_src/sim/builder.py

```diff
@@ -4403,6 +4436,97 @@ def finalize(self, device: Devicelike | None = None, requires_grad: bool = False
+                    if wp.types.type_is_vector(d):
+                        length = getattr(d, "_shape_", (1,))[0] or 1
+                        scalar_wp = getattr(d, "_wp_scalar_type_", wp.float32)
+                        scalar_np = wp.types.warp_type_to_np_dtype.get(scalar_wp, np.float32)
+                        try:
+                            base = np.array(
+                                default_val if default_val is not None else np.zeros(length, dtype=scalar_np),
+                                dtype=scalar_np,
+                            ).reshape(length)
+                        except Exception:
+                            base = np.zeros(length, dtype=scalar_np)
+                        arr = np.tile(base, (count, 1))
+                        # override values
+                        for idx, val in overrides.items():
+                            i = int(idx)
+                            if 0 <= i < count:
+                                try:
+                                    arr[i] = np.array(val, dtype=scalar_np).reshape(length)
+                                except Exception:
+                                    try:
+                                        arr[i] = np.array(list(val)[:length], dtype=scalar_np)
+                                    except Exception:
+                                        pass
+                        return wp.array(arr, dtype=d, requires_grad=requires_grad)
+
+                    # Scalar path
+                    scalar_np = wp.types.warp_type_to_np_dtype.get(d, np.float32)
+                    fill_val = default_val if default_val is not None else _default_for_dtype(d)
+                    arr = np.full(count, fill_val, dtype=scalar_np)
+                    for idx, val in overrides.items():
+                        i = int(idx)
+                        if 0 <= i < count:
+                            try:
+                                arr[i] = val
+                            except Exception:
+                                pass
+                    return wp.array(arr, dtype=d, requires_grad=requires_grad)
+
+                for var_name, spec in self.custom_properties.items():
+                    frequency = spec.get("frequency")
+                    if frequency not in {"joint", "joint_dof", "joint_coord", "body", "shape"}:
+                        continue
+
+                    # determine count by frequency
+                    if frequency == "body":
+                        count = m.body_count
+                    elif frequency == "shape":
+                        count = m.shape_count
+                    elif frequency == "joint":
+                        count = m.joint_count
+                    elif frequency == "joint_dof":
+                        count = m.joint_dof_count
+                    elif frequency == "joint_coord":
+                        count = m.joint_coord_count
+                    else:
+                        continue
```

Comment:
Can we make this frequency an `IntEnum` since there are only predefined possible options? Also this mapping from `IntEnum` can be extracted to a function, I think we have the same in `selection.py`. Both code paths should use the same enum/count handling. @nvlukasz for viz